plugins/catalog-backend/src/ingestion/LocationReaders.ts

```diff
@@ -0,0 +1,206 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getVoidLogger } from '@backstage/backend-common';
+import {
+  EntityPolicies,
+  EntityPolicy,
+  LocationSpec,
+} from '@backstage/catalog-model';
+import { Logger } from 'winston';
+import { AnnotateLocationEntityProcessor } from './processors/AnnotateLocationEntityProcessor';
+import { EntityPolicyProcessor } from './processors/EntityPolicyProcessor';
+import { FileReaderProcessor } from './processors/FileReaderProcessor';
+import { GithubReaderProcessor } from './processors/GithubReaderProcessor';
+import * as result from './processors/results';
+import {
+  LocationProcessor,
+  LocationProcessorDataResult,
+  LocationProcessorEntityResult,
+  LocationProcessorErrorResult,
+  LocationProcessorLocationResult,
+  LocationProcessorResult,
+  LocationProcessorSink,
+} from './processors/types';
+import { YamlProcessor } from './processors/YamlProcessor';
+import { LocationReader, ReadLocationResult } from './types';
+
+// The max amount of nesting depth of generated work items
+const MAX_DEPTH = 10;
+
+/**
+ * Implements the reading of a location through a series of processor tasks.
+ */
+export class LocationReaders implements LocationReader {
+  private readonly logger: Logger;
+  private readonly processors: LocationProcessor[];
+
+  static defaultProcessors(
+    entityPolicy: EntityPolicy = new EntityPolicies(),
+  ): LocationProcessor[] {
+    return [
+      new FileReaderProcessor(),
+      new GithubReaderProcessor(),
+      new YamlProcessor(),
+      new EntityPolicyProcessor(entityPolicy),
+      new AnnotateLocationEntityProcessor(),
+    ];
+  }
+
+  constructor(
+    logger: Logger = getVoidLogger(),
+    processors: LocationProcessor[] = LocationReaders.defaultProcessors(),
+  ) {
+    this.logger = logger;
+    this.processors = processors;
+  }
+
+  async read(location: LocationSpec): Promise<ReadLocationResult> {
+    const output: ReadLocationResult = { entities: [], errors: [] };
+    let items: LocationProcessorResult[] = [result.location(location, false)];
+
+    for (let depth = 0; depth < MAX_DEPTH; ++depth) {
+      const newItems: LocationProcessorResult[] = [];
+      const sink: LocationProcessorSink = i => newItems.push(i);
+
+      for (const item of items) {
+        if (item.type === 'location') {
+          await this.handleLocation(item, sink);
+        } else if (item.type === 'data') {
+          await this.handleData(item, sink);
+        } else if (item.type === 'entity') {
+          await this.handleEntity(item, sink, output);
+        } else if (item.type === 'error') {
+          await this.handleError(item, sink, output);
```

Comment:
I don't quite get the purpose of the split into these 4 handle methods tbh. As opposed to just letting the processor inspect the type and act accordingly.

Reply:
The "driver" contains the behaviors like "stop after first match" and similar. The processor is not aware of that coordination logic.

Reply:
Oh reading this again, you mean that there would be an extra `if` condition inside a single universal method in the processor? Well they still have different signatures, right

Reply:
Yeah I'm questioning why they need to have different signatures, instead of just a `handleItem` that returns zero or one item and optionally emits additional things

Reply:
What does returning zero items mean? What does returning one item mean? Is there a difference between returning one Buffer, and returning one Entity? Or returning one Entity and emitting one error? Etc. I am not sure how you'd stitch that up, man, sounds way more confusing

Reply:
Returning one item would mean the next processor in the chain picks it up. Returning zero means there's no further processing of it. Emitting items always sends them to the beginning of the chain. Throwing an error would then mean there's an error, and emitting errors I don't think would be a thing.

Reply:
attempted this in #1179 but actually am not fond of it